### DISCUSSION TITLE
Awesome new features: follow and libuv filesystem watcher

### DISCUSSION THREAD

**Discussion by cseickel** [nvim-neo-tree/neo-tree.nvim#68] [Announcements]
There are two new features for the filesystem source that I am really excited about but are opt-in right now.

The first is `follow_current_file` (#45 ) which will keep the tree focused on your current file as you navigate in ways outside of Neo-tree, such as with Telescope, Harpoon, or things like b#. Thanks to @ioreshnikov for suggesting this! I didn't realize how much I would like it until I started using it myself.

The second is `use_libuv_file_watcher` which will use nvim's libuv integration to watch the directories currently open in the tree and refresh whenever something is changed. Without this, we relied on autocmds like BufWritePost to detect changes. The problem with the autocmds is they only fire in response to you doing something within this nvim instance, and it completely misses all other ways the files might have been changed. Using libuv, we will catch all changes. Examples of what this would catch might be adding, deleting, or moving files from the terminal, or another vim instance, or checking out a different branch with git, etc. 

To enable both of these features, add these options to your config:

```lua
      require("neo-tree").setup({
        filesystem = {
          follow_current_file = true,
          use_libuv_file_watcher = true,
        }
      })
```

If you try out these features, please post back here to let me know what you think!

**Comment by gsalvatella** [comment]
I have just tested both features and they are fantastic. The follower is very handy to keep an eye on where you are, particularly if you're switching back and forth files in very dissimilar subfolders (e.g. kernel). The file watcher essentially removes the infamous manual `R` (reload) from NerdTree, and it becomes a blessing when switching branches and checking out from git.

**Comment by nyngwang** [comment]
Just realized that I haven't tried `use_libuv_file_watcher = true,`. Btw is this on the `main` branch?

**Comment by cseickel** [comment]
Yes, it has also been released to v1.x. I'm going to make this the default option in 2.0.

**Reply by nyngwang** [reply]
Good idea. I didn't know what's libuv before I read this thread, so I think many people will just set it to `false`.

**Comment by unphased** [comment]
Thanks for implicitly explaining that the config will work as we expect if we just give it a set of small overridden options like this. I have a similar question for every lua nvim plugin in existence since many consume huge config structures. I would prefer not to copy a snapshot of defaults presented in readmes since these tend to get very stale over time as plugins evolve. 

So, I'd like to ask, is this how most of them work? Is there perhaps some idiomatic lua code that is usually used to perform the object merge operation that I could try to grep for for a given plugin to see if I can safely just specify a subset of config that I want to set? Thanks. Really looking forward to enabling both of these features by the way. They are going to go a long way to make neovim superior to other tools.

**Reply by cseickel** [reply]
> So, I'd like to ask, is this how most of them work?

I'm pretty sure this is how ALL of them work. They will usually say so in the README file of the project. It's rare that there are any required config options without defaults.

**Comment by unphased** [comment]
Note for anyone in the same boat as me. the cursorline plugin `"yamatsum/nvim-cursorline"` is handy for word highlighting but it has "friendly" built in cursorline behavior which hides the cursorline on all but the active buffer, and that behavior applies "as intended" to the Neo-tree buffer (hiding the cursorline which is useful to show the active file per `follow_current_file`). 

Since I don't mind and want the native cursorline back just make sure if you use this plugin to setup `cursorline = { enable = false }` for it.
